pkg/operator/defragcontroller/defragcontroller.go

```diff
@@ -0,0 +1,188 @@
+		return fmt.Errorf("cluster is unhealthy: %s", memberHealth.Status())
+	}
+
+	var endpointStatus []*clientv3.StatusResponse
+	var leader *clientv3.StatusResponse
+	for _, member := range etcdMembers {
+		status, err := c.etcdClient.Status(ctx, member)
+		if err != nil {
+			return err
+		}
+		if leader == nil && status.Leader == member.ID {
+			leader = status
+			continue
+		}
+		endpointStatus = append(endpointStatus, status)
+	}
+
+	// leader last if possible
+	if leader != nil {
+		endpointStatus = append(endpointStatus, leader)
+	}
+
+	var errors []error
+	for _, status := range endpointStatus {
+		member, err := getMemberFromStatus(etcdMembers, status)
+		if err != nil {
+			errors = append(errors, err)
+		}
+
+		// Check each member's status which includes the db size on disk "DbSize" and the db size in use "DbSizeInUse"
+		// compare the % difference and if that difference is over the max diff threshold and also above the minimum
+		// db size we defrag the members state file. In the case where this command only partially completed controller
+		// can clean that up on the next sync. Having the db sizes slightly different is not a problem in itself.
+		if isEndpointFragmented(member, status) {
+			recorder.Eventf("DefragControllerDefragmentAttempt", "Attempting defrag on member: %s, dbSize: %d , dbInUse: %d, memberID %d, leader ID: %d", member.Name, status.DbSize, status.DbSizeInUse, member.ID, status.Leader)
+			if _, err := c.etcdClient.Defragment(ctx, member); err != nil {
+				errors = append(errors, err)
+			}
+			postDefragEndpointStatus, err := c.etcdClient.Status(ctx, member)
+			if err != nil {
+				errors = append(errors, err)
+			}
+
+			recorder.Eventf("DefragmentSuccess", "etcd member has been defragmented: %s, dbSize: %d, dbInUse: %d, member ID: %d, leader ID: %d", member.Name, postDefragEndpointStatus.DbSize, postDefragEndpointStatus.DbSizeInUse, member.ID, postDefragEndpointStatus.Leader)
+
+			//give cluster time to recover before we move to the next member.
+			if err := wait.Poll(
+				waitDuration,
+				timeoutDuration,
+				func() (bool, error) {
+					memberHealth, err := c.etcdClient.MemberHealth()
+					if err != nil {
+						return false, nil
+					}
+					if !etcdcli.IsClusterHealthy(memberHealth) {
+						klog.Warningf("cluster is unhealthy: %s", memberHealth.Status())
+						return false, nil
+					}
+					return true, nil
+				}); err != nil {
+				errors = append(errors, fmt.Errorf("timeout waiting for cluster to stabalize after defrag"))
+			}
+		}
+	}
+
+	return v1helpers.NewMultiLineAggregate(errors)
+}
+
+// isEndpointFragmented checks the status of all cluster members to ensure that no members have a fragmented store.
+// This can happen if the operator starts defrag of the cluster but then loses leader status and is rescheduled before
+// the operator can defrag all members.
+func isEndpointFragmented(member *etcdserverpb.Member, endpointStatus *clientv3.StatusResponse) bool {
```

Comment:
Just a suggestion: But if we don't pass the member (which we only use to log, and we can move that one up), we could nicely unit test this function.

Reply:
added full unit tests